group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -9607,6 +9607,151 @@ public void testOnConsumerGroupStateTransitionOnLoading() {
+        context.replay(RecordHelpers.newGroupEpochRecord(consumerGroupId, 10));
+
+        List<Record> records = new ArrayList<>();
+        context.groupMetadataManager.maybeDowngradeEmptyGroup(classicGroupId, records);
+        assertEquals(Collections.emptyList(), records);
+
+        // A classic group can't be downgraded.
+        context.groupMetadataManager.maybeDowngradeEmptyGroup(classicGroupId, records);
+        assertEquals(Collections.emptyList(), records);
+
+        // An empty consumer group can be upgraded.
+        context.groupMetadataManager.maybeDowngradeEmptyGroup(consumerGroupId, records);
+        assertEquals(Arrays.asList(
+            RecordHelpers.newTargetAssignmentEpochTombstoneRecord(consumerGroupId),
+            RecordHelpers.newGroupSubscriptionMetadataTombstoneRecord(consumerGroupId),
+            RecordHelpers.newGroupEpochTombstoneRecord(consumerGroupId)), records);
+        records.clear();
+
+        // A non-empty consumer group can't be downgraded.
+        ConsumerGroupMember.Builder memberBuilder = new ConsumerGroupMember.Builder(Uuid.randomUuid().toString());
+        context.replay(RecordHelpers.newMemberSubscriptionRecord(consumerGroupId, memberBuilder.build()));
+        context.groupMetadataManager.maybeDowngradeEmptyGroup(classicGroupId, records);
+        assertEquals(Collections.emptyList(), records);
+    }
+
+    @Test
+    public void testConsumerGroupHeartbeatWithEmptyClassicGroup() {
+        String classicGroupId = "classic-group-id";
+        String memberId = Uuid.randomUuid().toString();
+        MockPartitionAssignor assignor = new MockPartitionAssignor("range");
+        assignor.prepareGroupAssignment(new GroupAssignment(Collections.emptyMap()));
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .withAssignors(Collections.singletonList(assignor))
+            .build();
+        ClassicGroup classicGroup = new ClassicGroup(
+            new LogContext(),
+            classicGroupId,
+            EMPTY,
+            context.time,
+            context.metrics
+        );
+        context.replay(RecordHelpers.newGroupMetadataRecord(classicGroup, classicGroup.groupAssignment(), MetadataVersion.latestTesting()));
+
+        context.groupMetadataManager.getOrMaybeCreateClassicGroup(classicGroupId, false).transitionTo(PREPARING_REBALANCE);
+        assertThrows(GroupIdNotFoundException.class, () ->
+            context.consumerGroupHeartbeat(
+                new ConsumerGroupHeartbeatRequestData()
+                    .setGroupId(classicGroupId)
+                    .setMemberId(memberId)
+                    .setMemberEpoch(0)
+                    .setServerAssignor("range")
+                    .setRebalanceTimeoutMs(5000)
+                    .setSubscribedTopicNames(Arrays.asList("foo", "bar"))
+                    .setTopicPartitions(Collections.emptyList())));
```

Comment:
This feels a bit weird in this test as it does not align with its name. How about putting it in a separate test?